Tracking: fix inserter menu no results issue

retrofox opened the pull request:
#### Changes proposed in this Pull Request

This PR changes the way to track `no-results` search result from the `Inserter Menu`, replacing event delegation by using an [Slot](https://developer.wordpress.org/block-editor/developers/slotfills/), through of the, experimental, inserter menu extension.

Also, it implements a fallback implementation for the `7.8.1` core plugin version, since unfortunately, it has a bug. Basically, the `hasItems` parameter is always `true`, so we can't rely on it.
So, it checks the plugin version and applies the respective behavior according to this value. We should remove the fallback code once a new version of core-editor ships.

#### Markable changes

* Remove the event delegation tracker.
* Track same events using the `__experimentalInserterMenuExtension` component.
* It applies a temporary hacky fix to detect no-results search 

#### Testing instructions

1) Update your testing site with these changes (SB)
2) In your testing site, open the client dev console and set up the debug

```
> localStorage.setItem( 'debug', 'wpcom-block-editor*' )
```

A hard refresh is needed.

3) Open the Inserter Menu
4) Open the `Inserter Menu` selector
5) Generate event typing there
6) Confirm that it behaves as expected. It means when you get results, it should trigger a `wpcom_block_picker_search_term` event, as the screenshot shows below.

**blocks found** event
<img width="1463" alt="Screen Shot 2020-04-15 at 10 13 04 PM" src="https://user-images.githubusercontent.com/77539/79403892-8e5be400-7f66-11ea-95f5-790706ff3f87.png">

When there aren't blocks there, it also should trigger the `wpcom_block_picker_no_results` event too:

**blocks not found** event
<img width="1221" alt="Screen Shot 2020-04-15 at 10 13 58 PM" src="https://user-images.githubusercontent.com/77539/79403878-87cd6c80-7f66-11ea-8e51-64de9b20618f.png">

You can see what version is currently running in your site inspecting the debug line:

<img width="675" alt="Screen Shot 2020-04-16 at 8 15 16 AM" src="https://user-images.githubusercontent.com/77539/79449915-744bf100-7fba-11ea-9d6a-cf3a2cb9aa2b.png">


Fixes https://github.com/Automattic/wp-calypso/issues/41149

getdave commented on apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-navigation-menu-search-handler.js:
```diff
@@ -0,0 +1,112 @@
+		) {
+			tracksRecordEvent( 'wpcom_block_picker_no_results', eventProperties );
+		}
+	}, [ searchTerm ] );
```
I think you are missing deps here:

`plugingVersion` and `selectedBlock`

There could be more. The linter should tell you...

retrofox replied:
Not sure if we want to track plugin version, probably don't. Selected Block is right there, in the `eventProperties` object.

retrofox commented:
> ... but I guess you'll have to be conscious of checking the progress of the feature as it moves towards becoming stable in Core.

Yeah, right. This is an experimental feature but I hope it could get stable. Anyway, we need to pay attention to changes in it.
> I wanted to check - is the purpose of the PR to **fix a bug in the original** event delegation approach or is this **an entirely new approach**? If the later, what advantages does this approach provide over the original?

Both of them. Event delegation wasn't the issue but DOM inspection. In short, [we are realying on a DOM select query](https://github.com/Automattic/wp-calypso/blob/1d03056395f4bd7966c1cc9e3eee584eb037c106/apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-block-picker-search-term-handler.js#L52) to detect the `no-results` event. Since something changed in the DOM tree, this value is always true:

```ec6
const hasResults = document.querySelectorAll( '.block-editor-inserter__no-results' ).length === 0;
```

It means that every time that something changes here, it could affect the tracking data.

The slot approach has an API that fits better for our requeriments. The `__experimentalInserterMenuExtension` provides four arguments: `onSelect`, `onHover`, `filterValue`, `hasItems`. We don't have to care about DOM tree since everthing that we need is already there, even more.

> I guess my concern is we may be adding another possible tracking approach, which is halfway between delegation and full Redux which adds further complexity to the tracking system.

Event delegation should we replaced by a better one as soon as possible, imo. Relying to the DOM tree, that wasn't designed for it, sounds fragile. In fact, it's failing right now in production for this reason. 
I think the reason to change the approach is more than enough. My only concern is that this block is experimental, but despite this, I still think it's more stable than the DOM approach.

> 
> I could be wrong...

Me too :-)

> 
> ## Testing Results
> I'm not entirely clear from the testing instructions whether I need to test against different Gutenberg versions and how to do this. So I decided to go with "stable" and "edge" versions on Dotcom:

Sorry, my bad. Updating testing instructions.

> 
> I tested on a site stickered with Gutenberg Edge (running Gutenberg `7.9.0-rc.1`) and I encountered the following error:

This is perfect. I've updated the PR adding a debug line to make it easier to test.

> 
> <img alt="Screenshot 2020-04-16 at 10 10 30" width="1544" src="https://user-images.githubusercontent.com/444434/79437798-9483ac00-7fca-11ea-97e4-d6f7816a19f7.png">
> 
> I then tested on a non-stickered site with `7.8.1` and I got the expected results and no errors:
> 
> <img alt="Screenshot 2020-04-16 at 10 13 22" width="1619" src="https://user-images.githubusercontent.com/444434/79438098-ff34e780-7fca-11ea-83a6-c184594b1a3a.png">
> 

> I think there could be a problem with this on `rc-1`.

I don't think so. We need to check the string `7.8.1`. any different (higher) version should be ok.
 
> Thanks for your work here. Your knowledge of Gutenberg is impressive 👍

Not enough, tbh. Thanks for your words.

retrofox commented on apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-navigation-menu-search-handler.js:
```diff
@@ -0,0 +1,112 @@
+		selectedBlock: select( 'core/block-editor' ).getSelectedBlock(),
+	} ) );
+
+	const plugingVersion = window.wpcomGutenberg ? window.wpcomGutenberg.pluginVersion : null;
```
sorry, I don't get it. What references do you mean?

retrofox replied:
oh, the typo

retrofox commented on apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-navigation-menu-search-handler.js:
```diff
@@ -0,0 +1,112 @@
+
+	/*
+	 * Hacky temporal solution to detect no-results search result.
+	 * Unfortunately, __experimentalInserterMenuExtension has a bug
+	 * in the 7.8.1 core version.
+	 * The `hasItems` property is always true so it isn't possible
+	 * to rely on this value.
```
I haven't done a search for this, I just found the bug when I was development the code. Let me try it.

retrofox replied:
[found it](https://github.com/WordPress/gutenberg/pull/21138#discussion_r397723497)!!!

retrofox commented on apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-navigation-menu-search-handler.js:
```diff
@@ -0,0 +1,112 @@
+	 * The `hasItems` property is always true so it isn't possible
+	 * to rely on this value.
+	 *
+	 * @TODO: The following is a temporary solution and it should be removed
+	 *   and replaced by the usage of the `hasItems` property,
+	 *   once a new version of core is available in dotcom.
```
Nop. Creating a new one.

retrofox replied:
✅

simison commented on apps/wpcom-block-editor/src/wpcom/editor.js:
```diff
@@ -1,7 +1,25 @@
 import './features/reorder-block-categories';
 import './features/tracking';
+
+import InserterMenuTrackingEvent from './features/tracking/wpcom-navigation-menu-search-handler';
```
Shouldn't this be in `./features/tracking` and just imported here?

retrofox replied:
In that case, should we register the plugin there as well then, right?

mtias commented:
@getdave slots are a primary extensibility interface that we encourage as it's more idiomatic and easier to test maintain. We aim for slots to be semantic so that there are no weird couplings. For this use case, it makes sense also because I'd imagine you might also want to render helpful guides for certain queries (like "page" returning something about "here's how you can create a new page").

retrofox commented:
> @getdave slots are a primary extensibility interface that we encourage as it's more idiomatic and easier to test maintain.

This comment makes me more confident about the approach that we did since my concerning point was just that: how reliable is this feature to considering to put it working on prod.
